apps/condo/domains/property/components/panels/Builder/MapConstructor.spec.ts

```diff
@@ -0,0 +1,166 @@
+import { MapDebug } from './MapConstructor'
+
+const testSection = {
+    id: null,
+    minFloor: -5,
+    maxFloor: 10,
+    unitsOnFloor: 10,
+}
+
+const sectionName = () => {
+    return Math.random().toString()
+}
+const createBuildingMap = (sections: number) => {
+    const PropertyMap = new MapDebug()
+    for (let i = 0; i < sections; i++){
+        PropertyMap.mapAddSection({ ...testSection, name: sectionName() })
+    }
+    return PropertyMap
+}
+
+
+describe('Map constructor', () => {
+    describe('Service functions', () => {
+        describe('Select options for sections', () => {
+            it('Should correctly generate floor list for choosed section', () => {
+                const Building = createBuildingMap(5)
+                expect(Building.sections).toHaveLength(5)
+                expect(Building.getSectionFloorOptions(Building.sections[0].id)).toHaveLength(15)
+            })
+        })
+        describe('Select options for floors', () => {
+            it('should correctly generate sections list', () => {
+                const Building = createBuildingMap(14)
+                expect(Building.getSectionOptions()).toHaveLength(14)
+            })
+        })
+        describe('getUnitInfo by id', () => {
+            it('should correctly get unit position on section and floor', () => {
+                const Building = createBuildingMap(5)
+                const jsonMap = Building.getMap()
+                const unitFromMap = jsonMap.sections[2].floors[2].units[2]
+                const unit = Building.getUnitInfo(unitFromMap.id)
+                expect(unit.section).toEqual(jsonMap.sections[2].id)
+                expect(unit.floor).toEqual(jsonMap.sections[2].floors[2].id)
+            })
+        })
+        describe('geting floors range', () => {
+            it('should correctly get min and max floor from all sections', () => {
+                const PropertyMap = new MapDebug()
+                PropertyMap.mapAddSection({ id: null, minFloor: -2, maxFloor: 5, unitsOnFloor: 10, name: sectionName() })
+                PropertyMap.mapAddSection({ id: null, minFloor: -1, maxFloor: 10, unitsOnFloor: 10, name: sectionName() })
+                expect(PropertyMap.possibleFloors).toEqual([ 10, 9, 8, 7, 6, 5, 4, 3, 2, 1, -1, -2 ])
+            })
+        })
+    })
+
+    describe('Unit operations', () => {
+        describe('Add unit', () => {
+            it('unit should be correctly placed to map', () => {
+                const Building = createBuildingMap(10)
+                const jsonMap = Building.getMap()
+                const newUnit = { id: null, floor: jsonMap.sections[3].floors[3].id, section: jsonMap.sections[3].id, label: '1000' }
+                Building.mapAddUnit(newUnit)
+                Building.validate()
```

Comment:
Why we don't use validate during addUnit call? Is it means that we can add unit and get invalid data structure?

Reply:
It's better to protect developers from their mistakes and don't get them chance to have a mistake.

Reply:
I move validation to another section. Now there are 4 more tests to check validator and to check that operations with map and union produce valid json